File: parity/quantile_contract.mjs
// Inverse-CDF contract gate for the JS twin.
//
// `quantile` is documented as the inverse of `cdf`. Four mixture geometries
// broke that round-trip before the localized-bracket fix, and each one killed
// a different candidate repair:
//
//   1. tiny scale          — an ABSOLUTE stopping tolerance exceeds the whole
//                            bracket, so bisection stops after one halving;
//   2. far low-weight      — a mixture-moment bracket excludes the component
//      component            that owns the quantile, so it returns its own edge;
//   3. narrow inside wide  — a tolerance scaled to the MIXTURE sigma is too
//                            coarse for the narrow component;
//   4. astronomical        — bisecting the global span needs ~133 halvings
//      separation           against max_iter = 100.
//
// Mirrors tests/test_quantile_inverse_contract.py and
// rust/tests/quantile_contract.rs: the same numbers must hold in every port.
//
// Run: node parity/quantile_contract.mjs      (exits non-zero on any violation)

import { Dist } from "../docs/js/skaters/dist.mjs";

let failures = 0;

function check(name, ok, detail) {
  if (ok) {
    console.log(`ok   ${name}`);
  } else {
    console.log(`FAIL ${name}: ${detail}`);
    failures += 1;
  }
}

function roundtripErr(d, p) {
  return Math.abs(d.cdf(d.quantile(p)) - p);
}

// 1. tiny scale
{
  const d = new Dist([[1.0, 0.0, 1e-11]]);
  for (const p of [0.01, 0.5, 0.99]) {
    const e = roundtripErr(d, p);
    check(`tiny_scale p=${p}`, e < 1e-6, `roundtrip err ${e}`);
  }
  const q = d.quantile(0.99);
  check("tiny_scale value", Math.abs(q - 2.3263478740408408e-11) < 1e-13, `got ${q}`);
}

// 2. far low-weight component
{
  const d = new Dist([[0.9999, 0.0, 1.0], [1e-4, 1000.0, 1.0]]);
  const p = 1 - 1e-6;
  const q = d.quantile(p);
  check("far_component reaches it", q > 990.0, `stuck at bracket edge: ${q}`);
  check("far_component roundtrip", roundtripErr(d, p) < 1e-5, `err ${roundtripErr(d, p)}`);
}

// 3. narrow component inside a wide mixture
{
  const d = new Dist([[0.5, 0.0, 1e-11], [0.5, 1000.0, 1.0]]);
  const q = d.quantile(0.25);
  check("narrow_inside_wide roundtrip", roundtripErr(d, 0.25) < 1e-6, `err ${roundtripErr(d, 0.25)}`);
  check("narrow_inside_wide value", Math.abs(q) < 1e-9, `p=0.25 sits at the narrow median, got ${q}`);
}

// 4. astronomical separation beside a narrow component
for (const sep of [1e20, 1e40, 1e80]) {
  const d = new Dist([[0.5, 0.0, 1e-11], [0.5, sep, 1.0]]);
  for (const p of [0.25, 0.75]) {
    const e = roundtripErr(d, p);
    check(`astronomical sep=${sep} p=${p}`, e < 1e-6, `roundtrip err ${e}, q=${d.quantile(p)}`);
  }
}

// 5. deep tail: parade derives z through this call, so termination stays in
//    x-space. A probability-space exit blunts this from ~7.03 to ~6.0.
{
  const d = new Dist([[1.0, 0.0, 1.0]]);
  const z = d.quantile(1 - 1e-12);
  check("deep_tail diagnostic", Math.abs(z - 7.034481104929) < 1e-4, `got ${z}`);
}

// 6. unit scale must be untouched by the localization
{
  const d = new Dist([[0.6, 0.0, 1.0], [0.4, 3.0, 2.0]]);
  for (const p of [0.1, 0.5, 0.9]) {
    check(`unit_scale p=${p}`, roundtripErr(d, p) < 1e-6, `err ${roundtripErr(d, p)}`);
  }
}

if (failures > 0) {
  console.log(`\n${failures} inverse-CDF contract violation(s)`);
  process.exit(1);
}
console.log("\nall inverse-CDF contract checks passed");

File: rust/tests/quantile_contract.rs
//! `quantile` must invert `cdf` at every scale and for uneven mixtures.
//!
//! Rust mirror of `tests/test_quantile_inverse_contract.py`. Three regimes
//! broke the documented inverse-CDF contract:
//!
//! 1. tiny scales — the absolute tolerance exceeded the whole bracket, so
//!    bisection stopped after one halving (median of N(0, 1e-11) resolved
//!    to the ~0.003rd percentile);
//! 2. tiny-weight far components — the mixture-moment bracket excluded the
//!    answer, so bisection returned the bracket edge (~80 instead of
//!    1002.3263);
//! 3. astronomical separation beside a narrow component — bisecting the
//!    whole span needs ~133 halvings against max_iter = 100.
//!
//! The fix localizes before bisecting: sorted component +/- 8 sigma
//! endpoints are binary-searched for a gap containing p; the tolerance is
//! `tol` times the SMALLEST component sigma (which the narrow-inside-wide
//! case pins); termination stays in x-space (the deep-tail case pins this
//! — parade derives its z diagnostic through this function); and the
//! result is the `cdf >= p` endpoint rather than the midpoint.

use skaters_core::dist::Dist;

fn roundtrip_err(d: &Dist, p: f64) -> f64 {
    (d.cdf(d.quantile(p)) - p).abs()
}

#[test]
fn quantile_inverts_cdf_at_tiny_scale() {
    let d = Dist::new(vec![(1.0, 0.0, 1e-11)]);
    for p in [0.01, 0.5, 0.99] {
        assert!(roundtrip_err(&d, p) < 1e-6, "p={p} err={}", roundtrip_err(&d, p));
    }
    assert!((d.quantile(0.99) - 2.3263478740408408e-11).abs() < 1e-13);
}

#[test]
fn quantile_reaches_tiny_weight_far_component() {
    let d = Dist::new(vec![(0.9999, 0.0, 1.0), (1e-4, 1000.0, 1.0)]);
    let p = 1.0 - 1e-6;
    let q = d.quantile(p);
    assert!(q > 990.0, "quantile stuck at bracket edge: {q}");
    assert!(roundtrip_err(&d, p) < 1e-5);
}

#[test]
fn quantile_mixed_scales_narrow_inside_wide() {
    let d = Dist::new(vec![(0.5, 0.0, 1e-11), (0.5, 1000.0, 1.0)]);
    let q = d.quantile(0.25);
    assert!(roundtrip_err(&d, 0.25) < 1e-6, "q={q} cdf={}", d.cdf(q));
    assert!(q.abs() < 1e-9, "p=0.25 lives at the narrow component median, got {q}");
}

#[test]
fn quantile_astronomical_separation_with_narrow_component() {
    // Bisecting the GLOBAL span needs ~133 halvings to reach the narrow
    // component's scale, against max_iter = 100. Localizing to the endpoint
    // gap that contains p keeps the search where the tolerance can resolve.
    for sep in [1e20_f64, 1e40, 1e80] {
        let d = Dist::new(vec![(0.5, 0.0, 1e-11), (0.5, sep, 1.0)]);
        for p in [0.25_f64, 0.75] {
            let e = roundtrip_err(&d, p);
            assert!(e < 1e-6, "sep={sep} p={p} err={e} q={}", d.quantile(p));
        }
    }
}

#[test]
fn quantile_preserves_deep_tail_diagnostic() {
    // parade derives z = Phi^-1(pit) through this call; the clamp lands at
    // ~7.03 and a probability-space exit would blunt it to ~6.0 — that gap
    // is what this pins. Tolerance is 1e-4, not 1e-6: the Rust erf differs
    // from the Python reference by ~1.9e-6 in x at |z| ~ 7 (Python
    // 7.034481104929, Rust 7.034483039286). That difference is inherited
    // from erf and is IDENTICAL on pristine main — it is not introduced by
    // the bracket/tolerance change.
    let d = Dist::new(vec![(1.0, 0.0, 1.0)]);
    let z = d.quantile(1.0 - 1e-12);
    assert!((z - 7.034481104929).abs() < 1e-4, "deep-tail z was {z}");
    assert!(z > 6.5, "probability-space exit would have blunted this to ~6.0");
}

#[test]
fn quantile_unit_scale_unchanged() {
    let d = Dist::new(vec![(0.6, 0.0, 1.0), (0.4, 3.0, 2.0)]);
    for p in [0.1, 0.5, 0.9] {
        assert!(roundtrip_err(&d, p) < 1e-6);
    }
}
